### PR TITLE
Use failure instead of error_chain

### LINKED ISSUE #586: Use failure instead of error_chain
Thanks to Send/Sync issues, we are stranded on a fork of error_chain. In any case, it’s a painful and error-prone library to use. 

We should port to `failure`, which is the better replacement on the block.

And we should consider flattening out some of our error hierarchy, too.

### PR DESCRIPTION
**Pull request by grigoryk**
This is an early WIP to get feedback.

So far the only non-mechanical bit was conversion of `InvalidBinding`, since it wraps a `BindingError`.

Eventually fixes #586.

### CODE REVIEW

File: query-algebrizer/src/clauses/convert.rs [hunk 1]
```diff
@@ -27,10 +29,7 @@ use clauses::{
     ConjoiningClauses,
 };
 
-use errors::{
-    ErrorKind,
-    Result,
-};
+use errors::AlgebrizerError;
```
Comment (ncalexan) [inline query-algebrizer/src/clauses/convert.rs]: I'm not against declaring `Result<T> = std::result::Result<T, errors::AlgebrizerError>` to reduce the changes.  Or even implementing `bail!` in each crate -- it's easy.

The `Err(...)?` pattern is odd to me -- I prefer the `return Err(...)` expression.

Comment (rnewman) [reply query-algebrizer/src/clauses/convert.rs]: Using `?` adds an implicit `into` to do conversion of errors, which is a little more compact than `return Err(some_err.into())` or `return some_result.map_err(|e| e.into())`.

**Comment by grigoryk** [comment]
This is bulk of the work. Needs another cleanup pass, and rebasing on top of `master` - query-parser is gone, algebrizer changed a bit, and #736 is about to land.

**Comment by grigoryk** [comment]
Tests pass! I'm not super happy with how parts of this turned out, but it's not a good idea to let this fall behind the other work. We can iterate on the particulars once/if we hit problems along the line.

**Comment by grigoryk** [comment]
@ncalexan do you mind giving this a lookover before I have to rebase it again? :sob:

**Pull request closed without merging**